### PR TITLE
Adds configs to disable laggy hard deletes

### PR DESCRIPTION
**Pull request by MrsTonedOne**
## Changelog
:cl:
server: Added configs to disable laggy hard deletes once they lag the server too much.
admin: laggy hard deletes only output once per type path.
/:cl:

closes #58379

### CODE REVIEW

File: code/controllers/configuration/entries/general.dm
Comment (Rohesie) [inline code/controllers/configuration/entries/general.dm:518]:
```suggestion
/datum/config_entry/flag/disable_all_hard_deletes

```

Linter. This is for history/blame preservation.

File: code/controllers/configuration/entries/general.dm
Comment (Cyberboss) [inline code/controllers/configuration/entries/general.dm:518]: Linter pls

File: config/config.txt
Comment (Cyberboss) [inline config/config.txt:547]:
```suggestion
## Uncomment to disable hard deletes entirely, even things that explicitly request. This is not recommended unless you have a need for it during events or other high pop times where performance is key.
```

File: config/config.txt
Comment (Cyberboss) [inline config/config.txt:544]:
```suggestion
## Uncomment to disable hard deleting garbage collection failures. (Hard deleting GC failures causes lag in order to bring memory usage down and keep bugged objects from hanging around and causing bugs in other things.) You can safely enable this for performance on production if the lag spikes are too disruptive)
```

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Review by Cyberboss — APPROVED** [review]

File: code/controllers/configuration/entries/general.dm
Comment (Rohesie) [inline code/controllers/configuration/entries/general.dm:524]:
```suggestion
/datum/config_entry/flag/disable_all_hard_deletes

```
Protest noted.

File: config/config.txt
Comment (Rohesie) [inline config/config.txt:555]:
```suggestion
#DISABLE_ALL_HARD_DELETES

```

**Review by Rohesie — APPROVED** [review]

**Comment by LemonInTheDark** [comment]
I don't think this is a good step to take, at least not if it's used as you want to use it, which is across all servers all the time.

Hard deletes get fixed right now because they're a very present issue. They show up cleanly in the profiler, and they bug people like me.
If we remove the source of the bugging globally, they simply won't be fixed. That can be solved in part by testing #59612, but that won't catch anything near everything.

I'm scared of a situation where we let it go on too long because there's no "pain point" for it, and the amount of memory leaks begins to cause OOMs, at which point we are forced to fix a fuck ton of them.

From what I understand your issue with the current state of things is the amount of lag they generate at highpop, which while it makes a good motivator for this sort of thing, sucks ass.

Is there something we could do to limit hard deletion in situations like that without totally disabling it? maybe a config threshold for pop or "leaked vars" counts, or some other system? I'm very concerned about just turning it off totally

**Comment by Rohesie** [comment]
Maybe keeping track of the things that fail to hard delete through weak refs and a debug panel to access them would partly solve your issues with this? Maybe log those as well.

**Comment by LemonInTheDark** [comment]
That might help, but this still kinda rips out the hedonic treadmill I've been running on for a bit, and I'm not really sure what can replace that. Even if something is bad, if it doesn't effect you in some way you'll never fix it.

**Comment by Cyberboss** [comment]
I'm with Lemon on this one

**Comment by Timberpoes** [comment]
My thoughts on this are:

Hard deletes are runtime errors by any other name. They occur when we wish for something to be GCed but some other piece of code holds a reference to it. Instead of doing a stack trace, we instead tell byond to forcibly delete it - Nulling out those references in the process. Those null references may cause runtimes of their own, which helps track down weird behaviour relating to hard delete (for example, AI law datums had these kinds of issues).

I would argue a config to stop hard dels happening is the GC equivalent of `if(runtime_state) return` fixes that just serve to hide the error states away instead of bringing them to the surface to be dealt with.

I would much prefer this to be a var or proc or something in SSgarbage that admins can flick on and off at will. Sometimes we have to do a massive SDQL2 query to delete thousands of things and end up loading SSgarbage with thousands of hard dels. In the past I've used methods like emptying the bucket lists and queues to force the hard deleted objects to memory leak. I would have preferred to use a much less ghetto solution to "gracefully allowing" memory to leak.

**Comment by optimumtact** [comment]
see #maintainerbus

**Comment by optimumtact** [comment]
Every maintainer here is right, and you force this in against good reasonable technical sense.

**Review by Cyberboss — CHANGES_REQUESTED** [review]
This currently prevents disconnecting `/client`s via code.

**Comment by MrStonedOne** [comment]
So i forgot these vars edited on ssgarbage:

image

its the longest time it took to hard delete a single object, (sadly, no, the object type is not tracked) that was manuel at round 164627

this is sybil at 164628:

image

So, @optimumtact can you show me where sendmessage is taking 6 to 8 consecutive ticks to run once, or can you admit that that tangent was fucking stupid.

**Comment by LemonInTheDark** [comment]
We really should have had the conversation here from the start, makes it easier to track for future

QDEL logs from the sybil round, these are the only two above the max threshold
image
Based on the cost, I think it's reasonable to assume the issue lies with hotspots. Also based on the cost, I assume they're being held by either a datum or a turf somehow, but they're so intertwined I'd need to catch it on live to really solve it

**Comment by ninjanomnom** [comment]
If we'd like to reduce the impact on players from hard dels while at the same time still making them relevant as an incentive to fix things, it could be a good idea to only stop hard deleting types that have shown a major spike compared to other types. So for example at the start of the round everything is hard deleted like normal, if a single hard del takes [more than one tick/whatever duration] record the type and don't try to hard del that type anymore while at the same time giving a single warning in the admin channel that this has happened. You'll at worst get a single spike from any particular source and get a constant reminder if you ignore these things for too long.

**Comment by MrStonedOne** [comment]
If we'd like to reduce the impact on players from hard dels while at the same time still making them relevant as an incentive to fix things, it could be a good idea to only stop hard deleting types that have shown a major spike compared to other types. 

Working on this as a config

**Comment by MrStonedOne** [comment]
@ninjanomnom done

File: code/controllers/subsystem/garbage.dm
Comment (ninjanomnom) [inline code/controllers/subsystem/garbage.dm:277]: few of these, I assume you mean `|=`?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/garbage.dm
Comment (MrStonedOne) [reply code/controllers/subsystem/garbage.dm:277]: ugh, yes.

File: config/config.txt
Comment (LemonInTheDark) [inline config/config.txt:551]: the threshold needs to be enabled for the overrun limit to do anything doesn't it?

Comment (MrStonedOne) [reply config/config.txt:551]: yep. if threshold is set to 0, even the admin notices get disabled, and the if check for that overrun limit will never run.

the postpone code will still run.

Comment (LemonInTheDark) [reply config/config.txt:551]: It's a little unclear just from the config option, tho maybe that's better then giving people who won't read the code an understanding of it

Comment (MrStonedOne) [reply config/config.txt:551]: its 0.5 by default in code, hence the comment out in the config

**Comment by optimumtact** [comment]
@MrStonedOne I'm still waiting for you to inform me publicly here if you want this merged, or if you are accepting one or more of the compromise PR's.

Please comment so this PR can be moved forward.

**Pull request merged**
